From 66e6cd7b9f4d9fad863ff40db5ea6fc94a84d3b5 Mon Sep 17 00:00:00 2001
From: Tang Ding <tang.ding@tcl.com>
Date: Thu, 18 Mar 2021 13:53:13 +0000
Subject: [PATCH 32/49] FIx issue:
 https://issuetracker.google.com/issues/183024999

There's a performance issue during continually scroll listview when listview in Fling state .
Consider such scene as blew:
1.Listview holds a lots of data
2.mTouchMode in TOUCH_MODE_FLING state.
3.Touch Event delivery down event to listview , mTouchMode change to TOUCH_MODE_SCROLL , post checkFlyWheel(40ms timeout). move event make checkFlywheel.endFling()  delay  to execute  , before endFling() satisfied to execute up event occured, flingrunnable.start() will post a extra FlingRunnable in onTouchUp().

Finally Choreographer.CallbackQueue will get more and more FlingRunnable animation callback when keep repeating  above steps#3, Choreographer#doFrame#animation will  execute FlingRunnbale N times per frame , it will greater than vsync period cause jank (N means repeat steps#3 times).

Solution:
Remove unnecessary FlingRunnable before add FlingRunnable.

Change-Id: I2cbab9171ffbaad87a8908eef09f813cf89ad85a
---
 core/java/android/widget/AbsListView.java | 1 +
 1 file changed, 1 insertion(+)

diff --git a/core/java/android/widget/AbsListView.java b/core/java/android/widget/AbsListView.java
index 16e87f8bca9..dd9985ac5f4 100644
--- a/core/java/android/widget/AbsListView.java
+++ b/core/java/android/widget/AbsListView.java
@@ -4679,6 +4679,7 @@ public abstract class AbsListView extends AdapterView<ListAdapter> implements Te
                     0, Integer.MAX_VALUE, 0, Integer.MAX_VALUE);
             mTouchMode = TOUCH_MODE_FLING;
             mSuppressIdleStateChangeCall = false;
+            removeCallbacks(this);
             postOnAnimation(this);
 
             if (PROFILE_FLINGING) {
-- 
2.31.1

